### PR TITLE
fix: auto-recover from corrupted sandbox state

### LINKED ISSUE #5866: [bug]: Errors in post-request script persist
### Is there an existing issue for this?

- [x] I have searched existing issues and this bug hasn't been reported yet

### Platform

Desktop App

### Browser

Chrome

### Operating System

Windows

## What happened
When adding a post-request script which results in an error (e.g. `foo();`), after fixing the error or sometimes even after clearing the script, the test results pane says that there was an error. Sometimes closing an re-opening the request tab helps, most of the times only a restart of the app helps

## Console error messages
```
message: "cannot convert to object" name: "TypeError" stack: "    at <anonymous> (eval.js:1945)\n" [[Prototype]]: 
Object
```

## Steps to reproduce
- Add new request
- Leave the Echo URL as it is (`https://echo.hoppscotch.io`)
- Add a post-request script with an intentional error (e.g. `foo();`)
- Run request
- --> Error
- Clear post-request script or fix error
- Run request again
- --> Error persists

### Deployment Type

Hoppscotch Cloud

### Version

v2026.1.1

https://github.com/user-attachments/assets/82567b3f-d0f7-4c1e-b70b-417848d49afb

### PR DESCRIPTION
**Pull request by jamesgeorge007**
Closes #5866, #5810, FE-1128.

Persistent errors in the experimental scripting sandbox cause two user-visible bugs: 
- After fixing a broken pre/post-request script, the old error continues to surface because the singleton `FaradayCage` retains a corrupted state across executions.
- Requests with empty scripts report `script_fail` when the cached cage carries a stale bootstrap failure. This PR detects bootstrap/internal failures, resets the cage when needed, retries transparently on a fresh cage, and short-circuits empty scripts to avoid entering the sandbox entirely.

### What's changed

- **Bootstrap error detection via `captureHook`:** Extended the `captureHook` type from `{ capture?: () => void }` to `{ capture?: () => void; bootstrapError?: unknown }` in `scripting-modules.ts`. When `ctx.vm.callFunction(bootstrapIIFE)` returns an error, it is dumped and recorded on `captureHook.bootstrapError`. Both `pre-request/index.ts` and `test-runner/index.ts` check this flag as the primary cage reset signal.

- **Heuristic fallback via `isInfraError()`:** Defense-in-depth error classifier in `cage.ts` that catches errors escaping _before_ `callFunction()` is reached — specifically from `evalCode(bootstrapCode).unwrap()`, `createScriptingInputsObj()`, and `defineSandboxObject()`. Uses `instanceof Error` as the sole discriminator: infrastructure errors (QuickJSUnwrapError from `.unwrap()`, marshal failures, WASM init) are real `Error` instances, while user script errors from QuickJS's `dump()` are plain objects with `{ name, message, stack }` but are NOT `instanceof Error`. This structural difference makes the check reliable with no false positives.

- **Retry-on-bootstrap-error:** When a bootstrap error is detected, the cage is reset and the script is retried once on a fresh cage within the same execution. This eliminates the regression where fixing a broken script still fails on the _next immediate_ run before succeeding on the one after. Extracted `executePreRequestOnCage()` and `executeTestOnCage()` helper functions that return `E.Either | "retry"`. The caller retries once; if the retry also fails with a bootstrap error, surfaces `"sandbox initialisation error (persistent)"`.

- **Cage reset on bootstrap errors:** The reset decision is inlined at each call site as `bootstrapFailed || isInfraError(err)`. When `bootstrapFailed` is `true`, it short-circuits without inspecting the error. Otherwise, `isInfraError()` checks the error structure. When a reset is needed, callers invoke `resetCage()` which nulls `cagePromise`. User script errors (e.g., `ReferenceError: a is not defined`) do **not** trigger resets, preserving the singleton optimization from #5800.

- **`cachedCage` → `cagePromise` caching:** The singleton now caches the creation _promise_ rather than the resolved instance, which is the correct pattern for async singleton initialization. This prevents duplicate `FaradayCage.create()` calls if `acquireCage()` is called while a previous creation is still in-flight (e.g., after a cage reset, or on first use when two tabs fire requests before the cage is ready). Includes `.catch()` to null the promise on creation failure so subsequent retries work.

- **Empty script short-circuit:** `RequestRunner.ts` returns immediately (with identity envs/cookies and empty test results) when `stripModulePrefix(script).trim().length === 0`, avoiding WASM initialisation and preventing stale cage errors from surfacing as `script_fail` (#5810).

- **Structured console logging:** `RequestRunner.ts` error paths now use prefixed `console.error` (`[Pre-Request Script Error]`, `[Post-Request Script Error]`, `[Request Error]`) for dev-console visibility when debugging sandbox failures.

- **`SandboxTestResult` type:** Simplified from `TestResult & { tests: TestDescriptor } & { consoleEntries?; updatedCookies }` to a standalone type (`{ tests: TestDescriptor; envs: TestResult["envs"]; consoleEntries?: ConsoleEntry[]; updatedCookies: Cookie[] | null }`). The old intersection produced a contradictory `tests: TestDescriptor[] & TestDescriptor` — TypeScript allowed it structurally but every construction site always assigned a single `TestDescriptor`. The new type correctly reflects actual usage. Construction sites use `satisfies SandboxTestResult` for compile-time validation.

- **TaskEither contract compliance:** Both node experimental runners (`pre-request/experimental.ts`, `test-runner/experimental.ts`) now wrap the entire async body in `try/catch`, ensuring the returned `TaskEither` never rejects. The refactor replaces `pipe(TE.tryCatch(...))` with explicit `E.left`/`E.right` returns and a manual `try/catch`, preserving identical behavior while enabling the retry-on-bootstrap-error pattern. The inner helper functions format QuickJS error objects with the error `name` when available (e.g., `TypeError: cannot read property...`); the outer catch handles unexpected infrastructure throws. Web implementations (`web/pre-request/index.ts`, `web/test-runner/index.ts`) follow the same try-catch pattern for consistency.

- **Worker error handling:** Web worker catch blocks (`pre-request/worker.ts`, `test-runner/worker.ts`) now handle non-Error throws safely via `error instanceof Error ? error.message : String(error)` instead of the previous unsafe `(error as Error).message` cast. Web test-runner worker includes `updatedCookies: null` in the `SandboxTestResult` and uses `satisfies` for compile-time validation.

### Notes to reviewers

Try the reproduction steps mentioned in #5866 and similar patterns. Also, ensure to perform benchmarks to validate the previous performance optimisation strategies hold.

**Relationship to #5800:** This PR directly extends the singleton cage introduced in the memory leak fix (#5800). The key invariant preserved here: user script errors must _never_ trigger a cage reset — only bootstrap/internal failures (corrupted QuickJS context, WASM init failure, module setup errors) cause a reset. The singleton is preserved for all normal operations.

**Two-layer detection is intentional, not redundant.** `captureHook.bootstrapError` covers errors from `ctx.vm.callFunction(funcHandle)` in `scripting-modules.ts` (L476). `isInfraError()` covers errors that escape _before_ `callFunction()` — specifically `evalCode(bootstrapCode).unwrap()` (L417), `createScriptingInputsObj()` (L423), and `defineSandboxObject()` (L474). These throw during `FaradayCage.runCode()`'s module `def()` phase and propagate as `{ type: "error", err }` without ever setting `captureHook.bootstrapError`. An audit of the full error flow confirmed `isInfraError()` is reachable, not dead code.

**`instanceof Error` as the discriminator in `isInfraError()`.** This is intentional, not overbroad. In the FaradayCage/QuickJS pipeline, user script errors are always plain objects (from QuickJS's `dump()`), while infrastructure errors are always `Error` instances (QuickJSUnwrapError from `.unwrap()`, marshal failures, etc.). The two shapes never overlap. The primary caller path only sends `result.err` (plain objects) through `isInfraError()` when `captureHook.bootstrapError` is not set — the `instanceof Error` branch is a safety net for edge-case infrastructure throws.

**Test coverage:**

- `cage.spec.ts` — 5 unit tests for `isInfraError()` covering `instanceof Error` discrimination, QuickJSUnwrapError subclass detection, plain-object user error exclusions, and edge cases (null, string errors)
- `script-error-recovery.spec.ts` — 5 integration tests: 3 running real scripts through the full `FaradayCage` pipeline verifying error → recovery (matching the exact reproduction steps from #5866), plus 2 singleton retry tests that inject a corrupted cage via `_setCagePromiseForTesting()` and verify (a) bootstrap errors trigger transparent retry on a fresh cage, and (b) user script errors do not trigger the retry path
- All assertions validate actual function return values or real cage execution results — no mocked units-under-test or trivially-true checks
- `request.spec.ts` — 3 test expectations updated to include the `TypeError:` prefix in error messages, matching the new error name formatting in the script runners

### CODE REVIEW

File: packages/hoppscotch-common/src/helpers/RequestRunner.ts [hunk 1]
```diff
@@ -369,6 +370,17 @@ const delegatePreRequestScriptRunner = (
   const { preRequestScript } = request
 
   const cleanScript = stripModulePrefix(preRequestScript)
+
+  // Short-circuit empty scripts to avoid unnecessary WASM initialization
+  if (cleanScript.trim().length === 0) {
+    return Promise.resolve(
+      E.right({
+        updatedEnvs: envs,
+        updatedCookies: cookies,
```
[the remark this answers is not in the document]
Comment (jamesgeorge007) [reply packages/hoppscotch-common/src/helpers/RequestRunner.ts:378]: The `SandboxPreRequestResult` type defines `updatedRequest?: HoppRESTRequest` — it's an optional field, so omitting it is type-safe and semantically correct (implicit `undefined`). The consumer [`applyScriptRequestUpdates`](https://github.com/hoppscotch/hoppscotch/blob/next/packages/hoppscotch-common/src/helpers/experimental-sandbox-integration.ts#L31) handles absence with `if (!updatedRequest) { return originalRequest }`, which covers both `undefined` and `null`.

Adding `updatedRequest: undefined` explicitly would be redundant — TypeScript already enforces the optional contract, and `satisfies` would catch any future change making the field required. Keeping it implicit avoids noise and aligns with how the type was designed to be used.

File: packages/hoppscotch-js-sandbox/src/web/test-runner/index.ts [hunk 2]
```diff
@@ -57,77 +62,133 @@ const runPostRequestScriptWithFaradayCage = async (
...
-    })
-  } finally {
-    // FaradayCage relies on garbage collection for cleanup.
+    return retryResult
+  } catch (error) {
+    const name =
+      error instanceof Error && error.name ? `${error.name}: ` : ""
+    const message = error instanceof Error ? error.message : String(error)
+    return E.left(`Script execution failed: ${name}${message}`)
   }
```
Comment (jamesgeorge007) [inline packages/hoppscotch-js-sandbox/src/web/test-runner/index.ts:191]: Note on outer catch blocks: `FaradayCage.runCode()` catches internally and always returns `{ type: "error" }` — it never throws ([source](https://github.com/AndrewBastin/faraday-cage/blob/1f2bd58a65c1d9ba14592418b643a4ad31b1bd8d/lib/main.ts#L50-L124)). These outer catch blocks primarily guard against `acquireCage()` -> `FaradayCage.create()` rejection (WASM load failure, network error).
